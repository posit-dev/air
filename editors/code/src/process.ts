import * as cp from "child_process";

export interface CommandResult {
	type: "result";
	code: number | null;
	stdout: string;
	stderr: string;
}

export interface CommandError {
	type: "error";
	error: Error;
}

export function isResult(
	result: CommandResult | CommandError,
): result is CommandResult {
	return result.type == "result";
}

export function isError(
	result: CommandResult | CommandError,
): result is CommandError {
	return result.type == "error";
}

/**
 * Spawns a process and runs a command
 *
 * Collects stdout and stderr emitted along the way.
 *
 * @returns Returns a promise that results when the process exits or errors
 */
export async function runCommand(
	command: string,
	args?: readonly string[],
	options?: cp.SpawnOptionsWithoutStdio,
): Promise<CommandResult | CommandError> {
	return new Promise<CommandResult | CommandError>((resolve) => {
		let stdout = "";
		let stderr = "";

		// Use spawn instead of exec to avoid maxBufferExceeded error
		const p = cp.spawn(command, args, options);

		p.stdout.setEncoding("utf8");
		p.stdout.on("data", (data) => (stdout += data));

		p.stderr.setEncoding("utf8");
		p.stderr.on("data", (data) => (stderr += data));

		// `error` should fire before `close` in the event both fire.
		// `error` should also fire no matter what.
		p.on("error", (error) => {
			return resolve({
				type: "error",
				error: error,
			});
		});

		// Use `close` to wait for stdio streams to close too
		p.on("close", (code) => {
			return resolve({
				type: "result",
				code: code,
				stdout: stdout,
				stderr: stderr,
			});
		});
	});
}
