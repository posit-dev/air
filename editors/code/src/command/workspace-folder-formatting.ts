import * as vscode from "vscode";

import { Cmd, Ctx } from "../context";
import * as output from "../output";
import { isError, isResult, runCommand } from "../process";
import {
	saveAllDirtyWorkspaceTextDocuments,
	selectWorkspaceFolder,
} from "../workspace";

/**
 * Format a workspace folder
 *
 * # Tab saving
 *
 * Because we use the embedded Air CLI to perform the formatting, we force all
 * relevant tabs to be saved before proceeding. The Air CLI can't know about any
 * in-memory changes that may have occurred in "dirty" editors. Forcing them to
 * save causes any changes to be written to disk first, syncing VS Code with the
 * CLI.
 *
 * We don't force the tabs to be closed. We've seen that VS Code seems to sync
 * the LSP client and server well when performing a git branch switch (as long
 * as all files are saved beforehand), so we are hopeful it should also behave
 * well when we format with the CLI.
 *
 * # Forward propagation of settings
 *
 * If no `air.toml` is tied to a document, then typically in the LSP document
 * settings are forward propagated from the client to the server, and those are
 * used during formatting when that document is open. If an `air.toml` does
 * exist, those settings override document settings.
 *
 * Because we go through the Air CLI here, forward propagated document settings
 * are ignored, so even if you don't have an `air.toml`, you will get Air's
 * default indentation settings.
 *
 * We can't simply require an `air.toml` at the workspace folder root to avoid
 * confusion here, because you may instead have an `air.toml` at
 * `{root}/R/air.toml` that would handle any R files in your project. Instead,
 * we hope this isn't common enough to come up much in practice.
 */
export async function workspaceFolderFormatting(
	workspaceFolder: vscode.WorkspaceFolder,
	binaryPath: string,
) {
	const allSaved = await saveAllDirtyWorkspaceTextDocuments(workspaceFolder);
	if (!allSaved) {
		return;
	}

	const workspaceFolderPath = workspaceFolder.uri.fsPath;

	// i.e., `air format {workspaceFolderPath} --no-color`
	const args = ["format", workspaceFolderPath, "--no-color"];

	// This should not matter since the path is explicitly supplied, but better to be safe
	const options = {
		cwd: workspaceFolderPath,
	};

	// Resolves when the spawned process closes or errors
	const result = await runCommand(binaryPath, args, options);

	let anyErrors = false;

	if (isError(result)) {
		// Something went horribly wrong in the process spawning or shutdown process
		anyErrors = true;
		output.log(
			`Errors occurred while formatting the ${workspaceFolder.name} workspace folder.\n${result.error.message}`,
		);
	}

	if (isResult(result)) {
		if (result.code !== 0) {
			// Air was able to run and exit, but we had an error along the way
			output.log(
				`Errors occurred while formatting the ${workspaceFolder.name} workspace folder.\n${result.stderr}`,
			);
			anyErrors = true;
		}
	}

	if (anyErrors) {
		const answer = await vscode.window.showInformationMessage(
			`Errors occurred while formatting the ${workspaceFolder.name} workspace folder. View the logs?`,
			{ modal: true },
			"Yes",
			"No",
		);

		if (answer === "Yes") {
			output.show();
		}

		return;
	}

	vscode.window.showInformationMessage(
		`Successfully formatted the ${workspaceFolder.name} workspace folder.`,
	);
}

export function workspaceFolderFormattingCallback(ctx: Ctx): Cmd {
	return async () => {
		const workspaceFolder = await selectWorkspaceFolder();

		if (!workspaceFolder) {
			return;
		}

		const binaryPath = ctx.lsp.getBinaryPath();

		await workspaceFolderFormatting(workspaceFolder, binaryPath);
	};
}
