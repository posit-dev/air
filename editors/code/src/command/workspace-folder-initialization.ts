import * as vscode from "vscode";

import { Cmd, Ctx } from "../context";
import {
	saveAllDirtyWorkspaceTextDocuments,
	selectWorkspaceFolder,
} from "../workspace";
import { workspaceFolderFormatting } from "./workspace-folder-formatting";
import { fileExists } from "../utils";

/**
 * Initialize a workspace folder for use with Air
 *
 * - Creates an empty `air.toml` if neither `air.toml` nor `.air.toml` already exist
 * - Creates `extensions.json` recommending `Posit.air-vscode`
 * - Updates `settings.json` to format-on-save R and Quarto, and sets default formatters
 * - Updates `.Rbuildignore` to ignore `.vscode/` and `air.toml` if one existed
 *
 * Optionally, the user can request an immediate formatting of the workspace folder
 * after initialization
 */
export async function workspaceFolderInitialization(
	workspaceFolder: vscode.WorkspaceFolder,
	binaryPath: string,
) {
	const allSaved = await saveAllDirtyWorkspaceTextDocuments(workspaceFolder);
	if (!allSaved) {
		return;
	}

	await createAirToml(workspaceFolder);
	await createExtensionsJson(workspaceFolder);
	await updateSettingsJson(workspaceFolder);
	await updateRbuildignore(workspaceFolder);

	const message = `Successfully initialized the ${workspaceFolder.name} workspace folder.`;
	const shouldFormatItem = `Click to format the ${workspaceFolder.name} workspace folder`;

	const item = await vscode.window.showInformationMessage(
		message,
		shouldFormatItem,
	);

	if (item === shouldFormatItem) {
		await workspaceFolderFormatting(workspaceFolder, binaryPath);
	}
}

export function workspaceFolderInitializationCallback(ctx: Ctx): Cmd {
	return async () => {
		const workspaceFolder = await selectWorkspaceFolder();

		if (!workspaceFolder) {
			return;
		}

		const binaryPath = ctx.lsp.getBinaryPath();

		await workspaceFolderInitialization(workspaceFolder, binaryPath);
	};
}

async function createAirToml(
	workspaceFolder: vscode.WorkspaceFolder,
): Promise<void> {
	const airTomlUri = vscode.Uri.joinPath(workspaceFolder.uri, "air.toml");
	const dotAirTomlUri = vscode.Uri.joinPath(workspaceFolder.uri, ".air.toml");

	if ((await fileExists(airTomlUri)) || (await fileExists(dotAirTomlUri))) {
		return;
	}

	// Create an empty `air.toml`
	const content = new TextEncoder().encode("");
	await vscode.workspace.fs.writeFile(airTomlUri, content);
}

/**
 * Create `extensions.json` with Air recommended
 *
 * Unlike `settings.json`, there is no API for interacting with an
 * `extensions.json` file. To keep things simple, we never try and update an
 * existing one, we only create one if the user didn't have one already. We
 * don't really want to get into the json parsing business just for this.
 */
async function createExtensionsJson(
	workspaceFolder: vscode.WorkspaceFolder,
): Promise<void> {
	const vscodeUri = vscode.Uri.joinPath(workspaceFolder.uri, ".vscode");
	const extensionsUri = vscode.Uri.joinPath(vscodeUri, "extensions.json");

	if (await fileExists(extensionsUri)) {
		// Just bail if the user already has `extensions.json`, we don't
		// try to update an existing one
		return;
	}

	if (!(await fileExists(vscodeUri))) {
		await vscode.workspace.fs.createDirectory(vscodeUri);
	}

	let contents = `
{
    "recommendations": [
        "Posit.air-vscode"
    ]
}
	`;
	contents = contents.trim();
	contents = contents + "\n";

	const bytes = new TextEncoder().encode(contents);

	await vscode.workspace.fs.writeFile(extensionsUri, bytes);
}

/**
 * This seems to be the only way to update language specific settings for a
 * particular workspace folder in a way that:
 *
 * - Doesn't destroy existing `[r]` or `[quarto]` settings that we aren't
 *   updating (like `config.update("[r]", value)` naively does)
 * - Doesn't pull extra inherited global settings (like `config.get()` would do)
 *
 * The trick is to `inspect()` all of the `[r]` configuration specific to just
 * the `workspaceFolder` and bulk update it, retaining all old settings but
 * overriding the ones we care about updating.
 *
 * This does unfortunately drop comments in the `[r]` and `[quarto]` sections,
 * but we can't do better.
 *
 * It would be great if you could do `update("[r].editor.formatOnSave")` to
 * precisely target a single value of a single language, but you cannot.
 */
async function updateSettingsJson(
	workspaceFolder: vscode.WorkspaceFolder,
): Promise<void> {
	const config = vscode.workspace.getConfiguration(
		undefined,
		workspaceFolder,
	);

	const configR = config.inspect("[r]")?.workspaceFolderValue || {};
	await config.update(
		"[r]",
		{
			...configR,
			"editor.formatOnSave": true,
			"editor.defaultFormatter": "Posit.air-vscode",
		},
		vscode.ConfigurationTarget.WorkspaceFolder,
	);

	const configQuarto = config.inspect("[quarto]")?.workspaceFolderValue || {};
	await config.update(
		"[quarto]",
		{
			...configQuarto,
			"editor.formatOnSave": true,
			"editor.defaultFormatter": "quarto.quarto",
		},
		vscode.ConfigurationTarget.WorkspaceFolder,
	);
}

async function updateRbuildignore(
	workspaceFolder: vscode.WorkspaceFolder,
): Promise<void> {
	const rbuildignoreUri = vscode.Uri.joinPath(
		workspaceFolder.uri,
		".Rbuildignore",
	);

	if (!(await fileExists(rbuildignoreUri))) {
		// Do nothing if the user doesn't have one, i.e. an R project
		// or the rare R package without an `.Rbuildignore`
		return;
	}

	const buffer = await vscode.workspace.fs.readFile(rbuildignoreUri);
	const content = new TextDecoder().decode(buffer);

	const newline = content.includes("\r\n") ? "\r\n" : "\n";
	const lines = content.split(newline);

	if (lines.at(-1) === "") {
		// Drop final line if it's empty, i.e. we split on newlines and the
		// file ended with a newline. This avoids writing a blank line between
		// the existing content and our new lines.
		lines.pop();
	}

	let anyMissing = false;
	const patterns = ["^\\.vscode$", "^[.]?air[.]toml$"];

	for (const pattern of patterns) {
		const exists = lines.find((line) => line === pattern);

		if (!exists) {
			anyMissing = true;
			lines.push(pattern);
		}
	}

	if (anyMissing) {
		const content = lines.join(newline) + newline;
		const buffer = new TextEncoder().encode(content);
		await vscode.workspace.fs.writeFile(rbuildignoreUri, buffer);
	}
}
