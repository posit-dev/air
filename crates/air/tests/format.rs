mod utils;

use std::process::Command;

use crate::utils::command::get_air_bin;
use crate::utils::command::run;
use crate::utils::fixtures::path_fixtures;

#[test]
fn test_check_when_no_formatting_is_required() {
    let fixtures = path_fixtures();

    let path = fixtures.join("format").join("formatted.R");
    let path = path.to_str().unwrap();

    insta::assert_snapshot!(run(Command::new(get_air_bin())
        .arg("format")
        .arg(path)
        .arg("--check")));

    insta::assert_snapshot!(run(Command::new(get_air_bin())
        .arg("format")
        .arg(path)
        .arg("--check=interactive")));

    insta::assert_snapshot!(run(Command::new(get_air_bin())
        .arg("format")
        .arg(path)
        .arg("--check=github")));
}

#[test]
fn test_check_output_format() {
    let fixtures = path_fixtures();

    let path1 = fixtures.join("format").join("needs-formatting-1.R");
    let path1 = path1.to_str().unwrap();

    let path2 = fixtures.join("format").join("needs-formatting-2.R");
    let path2 = path2.to_str().unwrap();

    insta::assert_snapshot!(run(Command::new(get_air_bin())
        .arg("format")
        .arg(path1)
        .arg(path2)
        .arg("--check")));

    insta::assert_snapshot!(run(Command::new(get_air_bin())
        .arg("format")
        .arg(path1)
        .arg(path2)
        .arg("--check=interactive")));

    insta::assert_snapshot!(run(Command::new(get_air_bin())
        .arg("format")
        .arg(path1)
        .arg(path2)
        .arg("--check=github")));
}

#[test]
fn test_check_invalid_output_format() {
    insta::assert_snapshot!(run(Command::new(get_air_bin())
        .arg("format")
        .arg("path.R")
        .arg("--check=foo")));
}
