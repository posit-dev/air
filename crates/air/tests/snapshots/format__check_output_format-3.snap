---
source: crates/air/tests/format.rs
expression: "run(Command::new(get_air_bin()).arg(\"format\").arg(path1).arg(path2).arg(\"--check=github\"))"
---
success: false
exit_code: 1
----- stdout -----
::error file=fixtures/format/needs-formatting-1.R,line=1::Would reformat
::error file=fixtures/format/needs-formatting-2.R,line=1::Would reformat

----- stderr -----

----- args -----
format
/Users/davis/files/programming/air/crates/air/fixtures/format/needs-formatting-1.R
/Users/davis/files/programming/air/crates/air/fixtures/format/needs-formatting-2.R
--check=github
